test/performance/tiled_matmul.cpp

```diff
@@ -0,0 +1,280 @@
+
+    // Schedule the consumer
+    Var mmxi("mmxi"), mmyi("mmyi");
+    mm.in()
+        .tile(y, x, mmyi, mmxi, tile_y, tile_x)
+        .vectorize(mmyi)
+        .vectorize(mmxi);
+
+    Buffer<LhsInt8> a_buf(acc, row);
+    fill_buffer_a(a_buf, row, acc);
+    A.set(a_buf);
+
+    Buffer<RhsInt8> b_buf(4, col, acc / 4);
+    fill_buffer_b(b_buf, col, acc);
+    B.set(b_buf);
+
+    Buffer<int32_t> out(col, row);
+
+    Func result = mm.in();
+
+    // Uncomment to check the asm
+    //result.compile_to_llvm_assembly(Internal::get_test_tmp_dir() + "tiled_matmul.ll", {A, B}, target);
+    //result.compile_to_assembly(Internal::get_test_tmp_dir() + "tiled_matmul.s", {A, B}, target);
+
+    auto time = Tools::benchmark(20, 20, [&]() {
+        result.realize(out);
+    });
+    std::cout << "Exec time: " << time << "\n";
+
+    for (int j = 0; j < row; ++j) {
+        for (int i = 0; i < col; ++i) {
+            int32_t val = 0;
+            for (int k = 0; k < acc; ++k) {
+                val += a_buf(k, j) * b_buf(k % 4, i, k / 4);
+            }
+            if (val != out(i, j)) {
+                std::cerr << "Invalid result at " << i << ", " << j << "\n"
+                          << out(i, j) << " != " << val << "\n";
+                return false;
+            }
+        }
+    }
+    std::cout << "Success!\n";
+    return true;
+}
+
+auto matmul_ss = &matmul<int8_t, int8_t>;
+auto matmul_us = &matmul<uint8_t, int8_t>;
+auto matmul_su = &matmul<int8_t, uint8_t>;
+auto matmul_uu = &matmul<uint8_t, uint8_t>;
+
+bool equal_eps(float lhs, float rhs, float eps) {
+    return std::abs(lhs - rhs) < eps;
+}
+
+bool matmul_bf16(Halide::Target target) {
+    (void)target;
+
+    // lhs: 32x16, rhs: 16x32
+    const int row = 32;
+    const int col = 32;
+    const int acc = 16;
+
+    Var x("x"), y("y");
+    ImageParam A(BFloat(16), 2, "lhs");
+    ImageParam B(BFloat(16), 3, "rhs");
+
+    RDom r(0, acc, "acc");
+
+    Func mm("matmul");
+    mm(x, y) = cast<float>(0);
+    mm(x, y) += cast<float>(cast<float>(A(r.x, y))) * cast<float>(B(r.x % 2, x, r.x / 2));
+
+    int tile_x = 8;
+    int tile_y = 8;
+    int tile_r = 2;
+
+    Var rxi("rxi"), ryi("ryi");
+    RVar rri("rri"), rro("rro");
+
+    mm.compute_at(mm.in(), x)
+        .store_in(MemoryType::AMXTile)
+        .update()
+        .tile(x, y, rxi, ryi, tile_x, tile_y, TailStrategy::GuardWithIf)
+        .split(r.x, rro, rri, tile_r)
+        .reorder({rri, rxi, ryi, rro, x, y})
+        .atomic()
+        .vectorize(rri)
+        .vectorize(rxi)
+        .vectorize(ryi);
+
+    Var ixi("ixi"), iyi("iyi");
+    mm.compute_at(mm.in(), x)
+        .tile(x, y, ixi, iyi, tile_x, tile_y)
+        .vectorize(ixi)
+        .vectorize(iyi);
+
+    // schedule the consumer
+    Var mmxi("mmxi"), mmyi("mmyi");
+    mm.in()
+        .tile(x, y, mmxi, mmyi, tile_x, tile_y)
+        .vectorize(mmxi)
+        .vectorize(mmyi);
+
+    Func result = mm.in();
+    //result.print_loop_nest();
```

Comment:
Don't check in commented-out code (unless there is a comment explaining why, as is done elsewhere in this file)